File: packages/java/src/main/java/io/github/goldziher/benchmark/Benchmark.java
package io.github.goldziher.benchmark;

import io.github.goldziher.htmltomarkdown.HtmlToMarkdown;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.HashMap;
import java.util.Map;

/**
 * Benchmark utility for HTML-to-Markdown conversion performance testing.
 *
 * <p>This program measures the performance of the HTML-to-Markdown converter
 * on real HTML files with configurable iterations and warmup runs.
 *
 * @since 2.17.0
 */
public final class Benchmark {
    /**
     * Default number of benchmark iterations.
     */
    private static final int DEFAULT_ITERATIONS = 50;

    /**
     * Default number of warmup iterations.
     */
    private static final int DEFAULT_WARMUP = 3;

    /**
     * Default profiling frequency in Hz.
     */
    private static final int DEFAULT_PROFILING_FREQUENCY = 1000;

    /**
     * Conversion factor from nanoseconds to seconds.
     */
    private static final double NANOS_TO_SECONDS = 1_000_000_000.0;

    /**
     * Conversion factor from bytes to megabytes.
     */
    private static final double BYTES_TO_MB = 1024.0 * 1024.0;

    /**
     * Main entry point for the benchmark.
     *
     * @param args command-line arguments
     */
    public static void main(final String[] args) {
        String filePath = null;
        int iterations = DEFAULT_ITERATIONS;
        int warmup = DEFAULT_WARMUP;
        String format = "html";
        String scenario = "convert-default";

        for (int i = 0; i < args.length; i++) {
            switch (args[i]) {
                case "--file":
                    if (i + 1 < args.length) {
                        filePath = args[++i];
                    }
                    break;
                case "--iterations":
                    if (i + 1 < args.length) {
                        iterations = Integer.parseInt(args[++i]);
                    }
                    break;
                case "--warmup":
                    if (i + 1 < args.length) {
                        warmup = Integer.parseInt(args[++i]);
                    }
                    break;
                case "--format":
                    if (i + 1 < args.length) {
                        format = args[++i];
                    }
                    break;
                case "--scenario":
                    if (i + 1 < args.length) {
                        scenario = args[++i];
                    }
                    break;
            }
        }

        if (filePath == null) {
            System.err.println("Error: --file is required");
            System.exit(1);
        }

        if (!scenario.equals("convert-default")
                && !scenario.equals("metadata-default")) {
            System.err.println("Unsupported scenario: " + scenario);
            System.exit(1);
        }

        try {
            Path path = Paths.get(filePath);
            String html = Files.readString(path);

            String warmupEnv =
                System.getenv("HTML_TO_MARKDOWN_BENCH_WARMUP");
            if (warmupEnv != null && !warmupEnv.isBlank()) {
                try {
                    warmup = Integer.parseInt(warmupEnv);
                } catch (NumberFormatException ignored) {
                    warmup = DEFAULT_WARMUP;
                }
            }
            if (warmup < 0) {
                warmup = 0;
            }

            for (int i = 0; i < warmup; i++) {
                runScenario(html, scenario);
            }

            String profileOutput =
                System.getenv("HTML_TO_MARKDOWN_PROFILE_OUTPUT");
            if (profileOutput != null && !profileOutput.isBlank()) {
                int frequency = DEFAULT_PROFILING_FREQUENCY;
                String freqEnv =
                    System.getenv("HTML_TO_MARKDOWN_PROFILE_FREQUENCY");
                if (freqEnv != null && !freqEnv.isBlank()) {
                    try {
                        frequency = Integer.parseInt(freqEnv);
                    } catch (NumberFormatException ignored) {
                        frequency = DEFAULT_PROFILING_FREQUENCY;
                    }
                }
                HtmlToMarkdown.startProfiling(profileOutput, frequency);
            }

            long startNanos = System.nanoTime();
            for (int i = 0; i < iterations; i++) {
                runScenario(html, scenario);
            }
            long endNanos = System.nanoTime();

            if (profileOutput != null && !profileOutput.isBlank()) {
                HtmlToMarkdown.stopProfiling();
            }

            double elapsedSeconds = (endNanos - startNanos) / NANOS_TO_SECONDS;
            int bytesProcessed = html.getBytes().length * iterations;
            double opsPerSec = iterations / elapsedSeconds;
            double mbPerSec = (bytesProcessed / BYTES_TO_MB) / elapsedSeconds;

            Map<String, Object> result = new HashMap<>();
            result.put("language", "java");
            result.put("fixture", path.getFileName().toString());
            result.put("fixture_path", filePath);
            result.put("scenario", scenario);
            result.put("iterations", iterations);
            result.put("elapsed_seconds", elapsedSeconds);
            result.put("ops_per_sec", opsPerSec);
            result.put("mb_per_sec", mbPerSec);
            result.put("bytes_processed", bytesProcessed);

            System.out.println(toJson(result));

        } catch (IOException e) {
            System.err.println("Error reading file: " + e.getMessage());
            System.exit(1);
        }
    }

    /**
     * Convert a map to a JSON string representation.
     *
     * @param map the map to convert
     * @return JSON string representation
     */
    private static String toJson(final Map<String, Object> map) {
        StringBuilder sb = new StringBuilder("{");
        boolean first = true;
        for (Map.Entry<String, Object> entry : map.entrySet()) {
            if (!first) {
                sb.append(",");
            }
            first = false;
            sb.append("\"").append(entry.getKey()).append("\":");
            Object value = entry.getValue();
            if (value instanceof String) {
                sb.append("\"").append(value).append("\"");
            } else {
                sb.append(value);
            }
        }
        sb.append("}");
        return sb.toString();
    }

    /**
     * Run a benchmark scenario.
     *
     * @param html the HTML content to process
     * @param scenario the scenario name
     */
    private static void runScenario(final String html,
            final String scenario) {
        if ("metadata-default".equals(scenario)) {
            HtmlToMarkdown.convertWithMetadata(html);
        } else {
            HtmlToMarkdown.convert(html);
        }
    }

    /**
     * Private constructor to prevent instantiation.
     */
    private Benchmark() {
        throw new UnsupportedOperationException("Utility class");
    }
}

File: packages/go/v2/bin/benchmark.go
// Package main provides a CLI for benchmarking the Go bindings.
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"os"
	"path/filepath"
	"time"

	"github.com/Goldziher/html-to-markdown/packages/go/v2/htmltomarkdown"
)

type BenchmarkResult struct {
	Language       string  `json:"language"`
	Fixture        string  `json:"fixture"`
	FixturePath    string  `json:"fixture_path"`
	Scenario       string  `json:"scenario"`
	Iterations     int     `json:"iterations"`
	ElapsedSeconds float64 `json:"elapsed_seconds"`
	OpsPerSec      float64 `json:"ops_per_sec"`
	MBPerSec       float64 `json:"mb_per_sec"`
	BytesProcessed int     `json:"bytes_processed"`
}

func main() {
	filePath := flag.String("file", "", "Path to the HTML/HOCR fixture")
	iterations := flag.Int("iterations", 50, "Number of iterations")
	_ = flag.String("format", "html", "Fixture format (html or hocr)")
	scenario := flag.String("scenario", "convert-default", "Scenario to benchmark")
	flag.Parse()

	if *filePath == "" {
		fmt.Fprintln(os.Stderr, "Error: --file is required")
		os.Exit(1)
	}

	if *scenario != "convert-default" && *scenario != "metadata-default" {
		fmt.Fprintf(os.Stderr, "Unsupported scenario: %s\n", *scenario)
		os.Exit(1)
	}

	htmlBytes, err := os.ReadFile(*filePath)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error reading file: %v\n", err)
		os.Exit(1)
	}

	html := string(htmlBytes)

	profileOutput := os.Getenv("HTML_TO_MARKDOWN_PROFILE_OUTPUT")
	profileFrequency := os.Getenv("HTML_TO_MARKDOWN_PROFILE_FREQUENCY")
	profileOnce := os.Getenv("HTML_TO_MARKDOWN_PROFILE_ONCE")
	profileRepeat := os.Getenv("HTML_TO_MARKDOWN_PROFILE_REPEAT")
	if profileOutput != "" {
		_ = os.Unsetenv("HTML_TO_MARKDOWN_PROFILE_OUTPUT")
		_ = os.Unsetenv("HTML_TO_MARKDOWN_PROFILE_FREQUENCY")
		_ = os.Unsetenv("HTML_TO_MARKDOWN_PROFILE_ONCE")
		_ = os.Unsetenv("HTML_TO_MARKDOWN_PROFILE_REPEAT")
	}

	// Warmup (avoid profiling the warmup run).
	err = runScenario(html, *scenario)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Warmup conversion failed: %v\n", err)
		os.Exit(1)
	}

	if profileOutput != "" {
		_ = os.Setenv("HTML_TO_MARKDOWN_PROFILE_OUTPUT", profileOutput)
		if profileFrequency != "" {
			_ = os.Setenv("HTML_TO_MARKDOWN_PROFILE_FREQUENCY", profileFrequency)
		}
		if profileOnce != "" {
			_ = os.Setenv("HTML_TO_MARKDOWN_PROFILE_ONCE", profileOnce)
		}
		if profileRepeat != "" {
			_ = os.Setenv("HTML_TO_MARKDOWN_PROFILE_REPEAT", profileRepeat)
		}
	}

	// Benchmark
	start := time.Now()
	for i := 0; i < *iterations; i++ {
		err = runScenario(html, *scenario)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Conversion failed: %v\n", err)
			os.Exit(1)
		}
	}
	elapsed := time.Since(start).Seconds()

	bytesProcessed := len(htmlBytes) * (*iterations)
	opsPerSec := float64(*iterations) / elapsed
	mbPerSec := (float64(bytesProcessed) / (1024 * 1024)) / elapsed

	result := BenchmarkResult{
		Language:       "go",
		Fixture:        filepath.Base(*filePath),
		FixturePath:    *filePath,
		Scenario:       *scenario,
		Iterations:     *iterations,
		ElapsedSeconds: elapsed,
		OpsPerSec:      opsPerSec,
		MBPerSec:       mbPerSec,
		BytesProcessed: bytesProcessed,
	}

	encoder := json.NewEncoder(os.Stdout)
	if err := encoder.Encode(result); err != nil {
		fmt.Fprintf(os.Stderr, "Error encoding JSON: %v\n", err)
		os.Exit(1)
	}
}

func runScenario(html string, scenario string) error {
	switch scenario {
	case "metadata-default":
		_, err := htmltomarkdown.ConvertWithMetadata(html)
		return err
	default:
		_, err := htmltomarkdown.Convert(html)
		return err
	}
}
